$MirOS: ports/lang/pcc/patches/patch-cc_cc_cc_c,v 1.2 2008/10/28 12:00:47 tg Exp $

	• Support for -g0
	• String cleanup

--- cc/cc/cc.c.orig	Thu Jan  1 00:00:00 1970
+++ cc/cc/cc.c	Mon Nov 10 01:48:41 2008
@@ -128,7 +128,8 @@ char	*tmp3;
 char	*tmp4;
 char	*outfile, *ermfile;
 char *Bprefix(char *);
-char *copy(char *, int),*setsuf(char *, char);
+char *copy(char *, int, size_t *);
+char *setsuf(char *, char);
 int getsuf(char *);
 int main(int, char *[]);
 void error(char *, ...);
@@ -412,7 +413,10 @@ main(int argc, char *argv[])
 				break;
 
 			case 'g': /* create debug output */
-				gflag++;
+				if (argv[i][2] == '0')
+					gflag = 0;
+				else
+					gflag++;
 				break;
 
 			case 'i':
@@ -862,8 +866,9 @@ nocom:
 		}
 		if (outfile) {
 #ifdef MSLINKER
-			char *s = copy("/OUT:", strlen(outfile));
-			strcat(s, outfile);
+			size_t slen;
+			char *s = copy("/OUT:", strlen(outfile), &slen);
+			strlcat(s, outfile, slen);
 			av[j++] = s;
 #else
 			av[j++] = "-o";
@@ -922,12 +927,13 @@ nocom:
 		if (pthreads)
 			av[j++] = "-lpthread";
 		if (!nostdlib) {
+			size_t slen;
 #ifdef MSLINKER
-			char *s = copy("/LIBPATH:", strlen(pcclibdir));
+			char *s = copy("/LIBPATH:", strlen(pcclibdir), &slen);
 #else
-			char *s = copy("-L", strlen(pcclibdir));
+			char *s = copy("-L", strlen(pcclibdir), &slen);
 #endif
-			strcat(s, pcclibdir);
+			strlcat(s, pcclibdir, slen);
 			av[j++] = s;
 			if (pgflag) {
 				for (i = 0; libclibs_profile[i]; i++)
@@ -1042,6 +1048,7 @@ Bprefix(char *s)
 {
 	char *suffix;
 	char *str;
+	size_t len;
 
 	if (Bflag == NULL || s[0] != '/')
 		return s;
@@ -1050,8 +1057,8 @@ Bprefix(char *s)
 	if (suffix == NULL)
 		suffix = s;
 
-	str = copy(Bflag, strlen(suffix));
-	strcat(str, suffix);
+	str = copy(Bflag, strlen(suffix), &len);
+	strlcat(str, suffix, len);
 	return str;
 }
 
@@ -1073,7 +1080,7 @@ setsuf(char *s, char ch)
 {
 	char *p;
 
-	s = copy(basename(s), 2);
+	s = copy(basename(s), 2, NULL);
 	if ((p = strrchr(s, '.')) == NULL) {
 		p = s + strlen(s);
 		p[0] = '.';
@@ -1187,13 +1194,15 @@ callsys(char *f, char *v[])
  * Make a copy of string as, mallocing extra bytes in the string.
  */
 char *
-copy(char *s, int extra)
+copy(char *s, int extra, size_t *lenptr)
 {
 	int len = strlen(s)+1;
 	char *rv;
 
 	rv = ccmalloc(len+extra);
 	strlcpy(rv, s, len);
+	if (lenptr != NULL)
+		*lenptr = len + extra;
 	return rv;
 }
 
@@ -1225,7 +1234,7 @@ gettmp(void)
 		fprintf(stderr, "%s:\n", pathBuffer);
 		exit(8);
 	}
-	return copy(tempFilename, 0);
+	return copy(tempFilename, 0, NULL);
 }
 
 #else
@@ -1233,7 +1242,7 @@ gettmp(void)
 char *
 gettmp(void)
 {
-	char *sfn = copy("/tmp/ctm.XXXXXX", 0);
+	char *sfn = copy("/tmp/ctm.XXXXXX", 0, NULL);
 	int fd = -1;
 
 	if ((fd = mkstemp(sfn)) == -1) {
